tests-wasm/y-map.tests.js

```diff
@@ -76,21 +76,23 @@ export const testIterator = tc => {
     const x = d1.getMap('test')
 
     d1.transact(txn => {
-        x.set(txn, 'a', 1)
-        x.set(txn, 'b', 2)
-        x.set(txn, 'c', 3)
+        x.set('a', 1, txn)
+        x.set('b', 2, txn)
+        x.set('c', 3, txn)
     })
 
     let expected = {
         'a': 1,
         'b': 2,
         'c': 3
     }
-    for (let [key, value] of x.entries()) {
-        let v = expected[key]
-        t.compare(value, v)
-        delete expected[key]
-    }
+    d1.transact(txn => {
+        for (let [key, value] of x.entries(txn)) {
```

Comment:
Just a note: we can probably make `txn` optional in `.entries` as well.

Reply:
I was talking about this yesterday. The issue is that:

1. wasm_bindgen requires structs to be non-generic.
2. optional transactions are passed using reference `&Transaction`, but if it was not provided we need to instantiate them as non-reference `Transaction`
3. it means that for every iterator we need to have two types of structs defined: one for reference transaction and one for owned transaction.

This causes a lot of boilerplate to make. I can think about what to do with that.

Reply:
After few days of fighting with borrow checker (see commit: https://github.com/y-crdt/y-crdt/pull/199/commits/904839fd6cf3721287d116a30d916e0351aad56a) to make implicit transactions possible in this context, I realized that doing iterators in ywasm may not be possible - the reason is that every WASM object must be deallocated explicitly by calling `ywasmValue.free()` in JavaScript. If we don't do that iterator will hold the transaction forever, making other operations impossible.

Initially I wanted to use iterator [return](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#returnvalue) method to do the transaction cleanup. But there's a problem - it's only called when loop exits prematurelly (eg. due to `break;`). If loop ends normally, `iterator.return()` is never called, and since iterator is created by the loop itself under the hood, we also don't have reference that we could call free on.

I'm going to rewrite ywasm iterators to ordinary JS arrays to avoid memory leaks. Question is: **Should we leave changes in yrs iterators from commit https://github.com/y-crdt/y-crdt/pull/199/commits/904839fd6cf3721287d116a30d916e0351aad56a? They allow iterators to take ownership over passed transaction or just use transaction reference, which means that implicit iterators could potentially be used in other languages (y-py, y-rb), that can collect garbage after themselves**

/cc @dmonad @Waidhoferj @eliias

Reply:
@Horusiath could you not use the [Drop trait](https://doc.rust-lang.org/std/ops/trait.Drop.html)?

Reply:
@Horusiath do iterators by reference block operations inside the loop that need a mutable reference to that transaction? That is, does the iterator object hold onto the reference until it is garbage collected?

Reply:
@chanced The issue is not in the drop trait, but in a fact, that drop/free must be called manually in on an iterator created by yawsm to take an effect. Since iterator is created implicitly by the JS for loop, you never have a reference that you can call a `free()` on.

@Waidhoferj Yes, they do. Since iterator takes a readonly transaction reference, only readonly operations are allowed. Also since it never released, it means that you no longer be able to do anything mutating. I'd advice to check when pyo3 is calling Drop for you: does it have to be called explicitly or is it something automatic in specific cases.

Reply:
PyO3 uses a `PyRef` class, which is similar to an `Rc<RefCell<T>>` in its implementation. It calls the `Drop` trait automatically, but I believe it is batched with Python's gc pass, so there aren't true sequential consistency guarantees.

Reply:
What I did in ywasm eventually was to fallback from using iterators to return a full arrays/maps instead. They are created eagerly, so there's no need to keep on having a transaction for longer than it's necessary to build an array. This way all transaction references can be freed at the end of the method call.